### PR TITLE
add roommanager psrpc

### PR DESCRIPTION
**Pull request by paulwe**

### CODE REVIEW

File: pkg/routing/redisrouter.go
Comment (boks1971) [inline pkg/routing/redisrouter.go:378]: both keys could be available and this err could be `nil` wiping out the earlier error?

**Review by boks1971 — APPROVED** [review]
lgtm to @paulwe . But, my understanding of psrpc still leaves a lot to be desired 😄 . Some small comments, mostly due to my lack of understanding of things.

File: pkg/routing/localrouter.go
Comment (paulwe) [reply pkg/routing/localrouter.go:103]: we don't need connectionID anymore because this was only used to associate messages from media -> signaling with the response channels. i'm all but certain we could deprecate legacy participant keys but reconnect would fail if a user began a session using a new signal node then tried reconnecting from a legacy signal node with the old key format... idk what failure modes we tolerate

File: pkg/routing/localrouter.go
Comment (paulwe) [reply pkg/routing/localrouter.go:173]: local router only uses psrpc (using in memory channels). redis reuses the LocalRouter psrpc server (only using redis for a message bus) and reuses the LocalRouter client functions if psrpc is enabled

File: pkg/routing/redisrouter.go
Comment (paulwe) [reply pkg/routing/redisrouter.go:378]: oh you were intentionally wrapping both of them into the response. sure

Comment (boks1971) [reply pkg/routing/redisrouter.go:378]: yeah that was the intention, but I clearly had typos in there and not really wrapping the results 😞

File: pkg/routing/localrouter.go
Comment (boks1971) [reply pkg/routing/localrouter.go:103]: Got it.

On failure modes, would be good to have backwards compatibility and phased deprecation as projects using OSS and their own deploys should be able to upgrade without breaking compatibility. I need to spend more time thinking about the failure modes, but I am sure you have thought about compatibility. I am just stating the obvious 😄

Comment (paulwe) [reply pkg/routing/localrouter.go:103]: i feel like there's a maximum realistic amount of skew though. having a single deployment running 3 different versions seems like asking for bugs

Comment (boks1971) [reply pkg/routing/localrouter.go:103]: Can't agree more :-)

File: pkg/service/roommanager.go
Comment (davidzhao) [inline pkg/service/roommanager.go:666]: when this happens, does it mean the connection has severed? if so it'd be great to log here.

File: pkg/service/roommanager.go
Comment (davidzhao) [inline pkg/service/roommanager.go:679]: same comment as above.

File: pkg/service/rpc/egress.pb.go
Comment (davidzhao) [inline pkg/service/rpc/egress.pb.go:3]: side note: it'd be great to use CI to generate these so we don't have these version number diffs.

File: pkg/service/rpc/roommanager.proto
Comment (davidzhao) [inline pkg/service/rpc/roommanager.proto:17]: I know this PR is already getting pretty large.

I think these could actually expand into RoomService calls so we can return the right responses and error messages. That would solve another headache of mine :)

**Review by davidzhao — APPROVED** [review]
Really fantastic work here! Just a few small notes

File: pkg/service/rpc/egress.pb.go
Comment (paulwe) [reply pkg/service/rpc/egress.pb.go:3]: added but this is failing due to a permission error - https://github.com/livekit/livekit/actions/runs/4070633980/jobs/7011671528

File: pkg/service/roomservice.go
Comment (paulwe) [inline pkg/service/roomservice.go:337]: forwarding rpcs to psrpc methods is simpler

File: pkg/service/rpc/gen/room.proto
Comment (paulwe) [inline pkg/service/rpc/gen/room.proto:20]: psrpc got support for bulk register/unregister and statically typed topics

File: pkg/service/wire.go
Comment (paulwe) [inline pkg/service/wire.go:217]: rpc got retries to fix subscription/call races

**Review by paulwe — COMMENTED** [review]
i think this is ready but i'm going to hold this until the cloud changes are ready in case i missed something.

File: pkg/routing/localrouter.go
Comment (davidzhao) [inline pkg/routing/localrouter.go:237]: it'd be great to take in context.Context as first param, so that it'd be easier to pass additional values across for Cloud

File: pkg/routing/localrouter.go
Comment (davidzhao) [inline pkg/routing/localrouter.go:242]: what's calling these topic registration methods today?

File: pkg/routing/localrouter.go
Comment (paulwe) [reply pkg/routing/localrouter.go:237]: we receive the remote context in handlers - it's not committed but i added this to make auth session passing work for cloud based on the grpc metadata feature

File: pkg/routing/localrouter.go
Comment (paulwe) [reply pkg/routing/localrouter.go:242]: roommanager calls these during room creation/deletion and signaling session open/close similar to cloud

File: pkg/routing/localrouter.go
Comment (paulwe) [reply pkg/routing/localrouter.go]: this is in psrpc now but i'll add comments before i land that diff

**Review by davidzhao — APPROVED** [review]
lgtm!

**Pull request closed without merging**
